### PR TITLE
fixed IAM roles and code cleanup.

### PR DESCRIPTION
**Pull request by prudhvigodithi**
Signed-off-by: pgodithi <pgodithi@amazon.com>

### Description

- fixed IAM roles and code cleanup.
- Removed unstable ECR repo creation code.
- Fixed wildcard IAM policies.
- Added Required IAM policy conditions
- Added ECR public policy for agent nodes.
- Added Optional flag `agentAssumeRole` for agent's to assume cross account role.
- Added `job-import-plugin` and `workflow-multibranch` plugins.
- Updated README document.
- Added docker resources.
- Added jenkins `JENKIN_JAVA_OPTS` to docker-compose file.
- Added docker `maintainer` label.
- Minor version bump `2.332.3-lts`.
- Added agent node IAM roles unit tests.

### Issues Resolved
https://github.com/opensearch-project/opensearch-ci/issues/65

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by peterzhuamazon** [comment]
Does this include the ECR fix where assume role should happen on agent not main node?
Since @zelinh latest PR on fixing `TAG_LATEST` is still running on main node.

Thanks.

File: lib/compute/agent-node-config.ts [hunk 1]
```diff
@@ -6,108 +6,188 @@
...
+     const assumeRole = stack.node.tryGetContext('agentAssumeRole');
 
-  public readonly SSHEC2KeySecretId: string;
+     const key = new KeyPair(stack, 'AgentNode-KeyPair', {
+       name: 'AgentNodeKeyPair',
+       description: 'KeyPair used by Jenkins Main Node to SSH into Agent Nodes',
+     });
+     Tags.of(key)
+       .add('jenkins:credentials:type', 'sshUserPrivateKey');
+     const AgentNodeRole = new iam.Role(stack, 'OpenSearch-CI-AgentNodeRole', {
```
Comment (rishabh6788) [inline lib/compute/agent-node-config.ts:58]: import Role construct and just use that. 
e.g., `import { Role } from '@aws-cdk/aws-iam'`
          `new Role(x, y, z);`

File: lib/compute/agent-node-config.ts [hunk 2]
```diff
@@ -6,108 +6,188 @@
...
+       assumedBy: new iam.ServicePrincipal('ec2.amazonaws.com'),
```
Comment (rishabh6788) [inline lib/compute/agent-node-config.ts:59]: same as mentioned about `Role` construct

File: lib/compute/agent-node-config.ts [hunk 3]
```diff
@@ -6,108 +6,188 @@
...
-        name: 'Name',
-        value: `jenkinsAgentNode-${config.workerLabelString}`,
-      }],
-      tenancy: 'Default',
-      type: config.instanceType,
-      useEphemeralDevices: false,
-    };
-  }
-}
+     const ecrManagedPolicy = new iam.ManagedPolicy(stack, 'OpenSearch-CI-AgentNodePolicy', {
```
Comment (rishabh6788) [inline lib/compute/agent-node-config.ts:65]: needs to same format as mentioned above about `Role`

File: lib/compute/agent-node-config.ts [hunk 4]
```diff
@@ -6,108 +6,188 @@
...
+       description: 'Jenkins agents Node Policy',
+       managedPolicyName: 'OpenSearch-CI-AgentNodePolicy',
+       statements: [
+         new iam.PolicyStatement({
```
Comment (rishabh6788) [inline lib/compute/agent-node-config.ts:69]: needs correction here

File: lib/compute/agent-node-config.ts [hunk 5]
```diff
@@ -6,108 +6,188 @@
...
+           effect: iam.Effect.ALLOW,
```
Comment (rishabh6788) [inline lib/compute/agent-node-config.ts:70]: here too

File: lib/compute/agent-node-config.ts [hunk 6]
```diff
@@ -6,108 +6,188 @@
...
+              'aws:RequestedRegion': this.STACKREGION,
+              'aws:PrincipalAccount': [this.ACCOUNT],
+            },
+          },
+         }),
+       ],
+       roles: [AgentNodeRole],
+     });
+     ecrManagedPolicy.addStatements(
+      new iam.PolicyStatement({
```
Comment (rishabh6788) [inline lib/compute/agent-node-config.ts:97]: here too

File: lib/compute/agent-node-config.ts [hunk 7]
```diff
@@ -6,108 +6,188 @@
...
+        conditions: {
+          'StringEquals': {
+            'aws:RequestedRegion': this.STACKREGION,
+            'aws:PrincipalAccount': [this.ACCOUNT],
+          },
+        },
+      }),
+    );
+    if (assumeRole !== undefined) {
+      console.log('Adding agent agentAssumeRole is  👉', agentAssumeRole.valueAsString);
```
Comment (rishabh6788) [inline lib/compute/agent-node-config.ts:112]: please do not use `console.log` statements.

File: lib/compute/agent-node-config.ts [hunk 8]
```diff
@@ -6,108 +6,188 @@
...
+            'aws:PrincipalAccount': [this.ACCOUNT],
+          },
+        },
+      }),
+    );
+    if (assumeRole !== undefined) {
+      console.log('Adding agent agentAssumeRole is  👉', agentAssumeRole.valueAsString);  
+      // policy to allow assume role AssumeRole
+      AgentNodeRole.addToPolicy(
+        new iam.PolicyStatement({
```
Comment (rishabh6788) [inline lib/compute/agent-node-config.ts:115]: All the places where `iam` module is used, please correct it to first import the relevant constructs and classes and then use the same without using `iam.`

**Comment by rishabh6788** [comment]
Added a few comments. 
Please also add unit tests to cover the new resources being created for agent node.

**Comment by prudhvigodithi** [comment]
Hey @rishabh6788 I have pushed your suggestions and also added some unit tests for agents. Please check

**Review by rishabh6788 — APPROVED** [review]
LGTM

**Pull request merged**